content/blog/2019-09-25-floats-static-arrays.md

```diff
@@ -0,0 +1,67 @@
+```
+Arrays can be written to freely in any Bril block.  As a result, array values are inherently stateful and should be treated by users carefully.  Writing values of the wrong type, however, is expected to not be permitted.
+
+### Implementation
+
+Floating points and fixed-length arrays have been implemented as described, and can be reasoned about by the `bril2json` parser, the `Brili` interpreter, and the `bril2txt` translator.  A subset of TypeScript can be written to Bril-recognized JSON by the `ts2bril` compiler.  While these conversions work generally as described above, there are some features which bear special notice.
+
+The `Brili` interpreter supports integers, doubles, and floating-points all as different objects.  Integers are represented as Javascript [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) objects.  This choice of representation means that `Brili` does not correctly reason about integer overflow -- while this should be implemented in the future, I have not observed a need for it in the current Bril projects.
+
+Doubles can be represented simply as `number`, while floats are derived with the [Math.fround](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/fround) function.  Note that `fround` is only applied _after_ each floating-point operation; this is permitted due to the [IEEE-754 restrictions](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html) on floating point operation precision loss (or lack of).  Extending Bril to support other mathematical operations (such as `pow` or `sqrt`), may require that this implementation choice be updated.
+
+TypeScript represents every number as a `double` by default, thus operations in TypeScript are translated to Bril as floating-point operations.  The TypeScript `bigint` type can be used to specially generate integers in Bril.  Operations between TypeScript `bigint` values are not supported at the time of this writing; however, this should be changed shortly.  Floating points cannot be easily represented by TypeScript code; however, I will look into using the `fround` function as a mechanism for generating Bril `float` values.
+
+Fixed-length arrays are reasoned about naively by the `Brili` interpreter; in particular, it makes no attempt to check array length before indexing.  `Brili` arrays are filled with 0s by default when initialized.
+
+`ts2bril` can compile TypeScript arrays of consistent types to a `Bril`-like language.  This language differs from Bril only in that fixed-length arrays may have a size of `-1`, indicating that the size of the array is not known.  I am planning to explore ways to fix this issue; as TypeScript does not track array length and the Bril semantics for arrays is extremely limited, it should be very straightforward to take a second pass after compiling to replace each instance of unknown array length with the correct size.
```

Comment:
To clarify, does the compiler _always_ emit -1 for the array size? Or are there circumstances when it can generate a real size?

I would actually posit that a second pass to deduce array sizes might actually be very hard to do. Especially if the array actually grows and shrinks, which TypeScript arrays can do!

So maybe the sensible thing would be to compile TypeScript tuples, which have fixed sizes, to Bril arrays. Or to add variable-length arrays to Bril.